### PR TITLE
Datapoint extrapolation

### PR DESCRIPTION
**Pull request by shorowit**
## Pull Request Description

Addresses #1871. Updates datapoint extrapolation per RESNET MINHERS Addendum 82.

## Checklist

Not all may apply:

- [x] Schematron validator (`EPvalidator.xml`) has been updated
- [x] Sample files have been added/updated (`openstudio tasks.rb update_hpxmls`)
- [x] Tests have been added/updated (e.g., `HPXMLtoOpenStudio/tests/test*.rb` and/or `workflow/tests/test*.rb`)
- [x] Documentation has been updated
- [x] Changelog has been updated
- [x] `openstudio tasks.rb update_measures` has been run
- [x] No unexpected changes to simulation results of sample files

### CODE REVIEW

**Comment by shorowit** [comment]
@yzhou601 This PR is ready, other than the one test failure:
```base-hvac-air-to-air-heat-pump-var-speed-max-power-ratio-schedule-two-systems.xml: [Peak Load: Cooling: Delivered (kBtu/hr), Annual] 1x=228.96999999999997, 10x=195.301, abs_delta_tol=0.2, abs_frac_tol=0.05```

File: HPXMLtoOpenStudio/resources/hvac.rb [hunk 1]
```diff
@@ -4182,7 +4182,8 @@ def self.add_variable_speed_power_ems_program(runner, model, air_loop_unitary, c
         program.addLine("  If #{clg_tot_sensor.name} > 0")
         program.addLine("    Set #{cap_multiplier} = #{clg_sens_sensor.name} / #{clg_tot_sensor.name}")
         program.addLine('  Else')
-        program.addLine("    Set #{cap_multiplier} = 0.0")
```
Comment (yzhou601) [inline HPXMLtoOpenStudio/resources/hvac.rb:4125]: The spikes happened when the point is the first hour in a cooling cycle, and the SHR was set to 0.0 here, which confused the EMS to execute the max power cutoff (thought that the load is greater than available capacity)

**Comment by yzhou601** [comment]
> @yzhou601 This PR is ready, other than the one test failure: `base-hvac-air-to-air-heat-pump-var-speed-max-power-ratio-schedule-two-systems.xml: [Peak Load: Cooling: Delivered (kBtu/hr), Annual] 1x=228.96999999999997, 10x=195.301, abs_delta_tol=0.2, abs_frac_tol=0.05`

Should be fixed in the last commit, let's give it a try and I'll review this PR after CI gets green.

File: HPXMLtoOpenStudio/resources/hvac.rb (hunk 1 above)
Comment (shorowit) [reply HPXMLtoOpenStudio/resources/hvac.rb:4125]: Pulled your commit into a separate PR so we can see the change by itself: https://github.com/NREL/OpenStudio-HPXML/pull/1909

File: HPXMLtoOpenStudio/resources/hvac.rb [hunk 2]
```diff
@@ -2981,32 +3016,52 @@ def self.interpolate_to_odb_table_point(detailed_performance_data, capacity_desc
...
 
+    # Linearly extrapolate
     slope = (right_dp.send(property) - left_dp.send(property)) / (right_odb - left_odb)
     val = (target_odb - left_odb) * slope + left_dp.send(property)
+
+    if mode == :clg && target_odb < 82
+      # Ensure no less than 50% of the value at 82F
+      # FIXME: Need interpretation from RESNET on how this case should be handled
+      dp_82F = data.find { |dp| dp.outdoor_temperature == 82.0 }
+      val = [val, 0.5 * dp_82F.send(property)].max
```
[the remark this answers is not in the document]
Comment (yzhou601) [reply HPXMLtoOpenStudio/resources/hvac.rb:3027]: There're some complications for different cases. The resnet rules are in extrapolation method, so if the user provides a powerAt60F:
1. If the powerAt60F< 0.5 * powerAt82F, the `extrapolate_data_point_to_odb` is not called for 60F (since it's in user_odbs not target_odb) and it is not capped. And if the gross input power becomes negative at an ODB > 60F(`target_odb > 60` in above comment), that value will be capped, and the interpolation between that odb and 60F will be below 0.5 * powerAt82F. 
It will be like:
image

2. If the powerAt60F > 0.5 * powerAt82F, the `extrapolate_data_point_to_odb` is not called for 60F (since it's in user_odbs not target_odb) and it doesn't need the cap. If the gross input power becomes negative at an ODB > 60F(`target_odb > 60` in above comment), the there'll be a dip at that ODB being capped at 0.5 * powerAt82F. Something like this: 
image

3. If the powerAt60F< 0.5 * powerAt82F, and the gross input power becomes negative at an ODB < 60F, the 60F user provided data will be used, then:

image

4. If the powerAt60F> 0.5 * powerAt82F, and the gross input power becomes negative at an ODB < 60F, the 60F user provided data will be used, doesn't need cap, and everything good.
image

So to summarize, the user provided 60F data will be always be used in current implementation, but RESNET rules are not always forced. To avoid the unexpected results, I'm inclined to check the user provided data to be compliant with RESNET rule, if not, throw a warning.

**Review by yzhou601 — COMMENTED** [review]
This is nice work to add the RESNET approach and clean up some old variables. A few comments.

File: HPXMLtoOpenStudio/resources/hvac.rb [hunk 3]
```diff
@@ -2849,52 +2850,74 @@ def self.process_neep_detailed_performance(hvac_system, mode, max_rated_fan_cfm,
...
     end
-    # convert net to gross
+
     data_array.each_with_index do |data, speed|
       data.each do |dp|
-        this_cfm = UnitConversions.convert(dp.capacity, 'Btu/hr', 'ton') * cfm_per_ton[speed]
-        fan_ratio = this_cfm / max_rated_fan_cfm
-        fan_power = calculate_fan_power_from_curve(hvac_ap.fan_power_rated * max_rated_fan_cfm, fan_ratio, hvac_system)
-        dp.gross_capacity, dp.gross_efficiency_cop = convert_net_to_gross_capacity_cop(dp.capacity, fan_power, mode, dp.efficiency_cop)
+        convert_data_point_net_to_gross(dp, mode, hvac_system, cfm_per_ton[speed], max_rated_fan_cfm)
```
[the remark this answers is not in the document]
Comment (shorowit) [reply HPXMLtoOpenStudio/resources/hvac.rb:2856]: It's needed because we still identify min/max ODB temperatures in the `extrapolate_data_points` method using gross values. I will add a comment explaining this.

File: HPXMLtoOpenStudio/resources/hvac.rb [hunk 4]
```diff
@@ -2849,52 +2850,74 @@ def self.process_neep_detailed_performance(hvac_system, mode, max_rated_fan_cfm,
...
-    extrapolate_data_points(data_array, mode, compressor_lockout_temp, weather_temp)
-    add_data_point_adaptive_step_size(data_array, mode)
+
+    extrapolate_data_points(data_array, mode, hp_min_temp, weather_temp, hvac_system, cfm_per_ton, max_rated_fan_cfm)
     correct_ft_cap_eir(data_array, mode)
   end
 
+  # TODO
+  #
+  # @param dp [TODO] The detailed performance data point of interest
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/hvac.rb:2866]: I should address these TODOs

File: HPXMLtoOpenStudio/resources/hvac.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (shorowit) [reply HPXMLtoOpenStudio/resources/hvac.rb:2856]: I'll see if the data_array loop that converts net to gross can be moved inside the `extrapolate_data_points` method.

File: HPXMLtoOpenStudio/resources/hvac.rb [hunk 5]
```diff
@@ -2905,11 +2928,17 @@ def self.extrapolate_data_points(data_array, mode, compressor_lockout_temp, weat
...
+        new_dp.capacity = extrapolate_data_point_to_odb(data, mode, capacity_description, target_odb, :capacity)
+        new_dp.input_power = extrapolate_data_point_to_odb(data, mode, capacity_description, target_odb, :input_power)
+        new_dp.efficiency_cop = new_dp.capacity / new_dp.input_power
+        convert_data_point_net_to_gross(new_dp, mode, hvac_system, cfm_per_ton[speed], max_rated_fan_cfm)
+
         data << new_dp
       end
     end
+
+    add_data_point_adaptive_step_size(data_array, mode, hvac_system, cfm_per_ton, max_rated_fan_cfm)
```
[the remark this answers is not in the document]
Comment (shorowit) [reply HPXMLtoOpenStudio/resources/hvac.rb:2941]: As we discussed, it is still needed. I am updating the description of the method a bit in the next commit.

File: HPXMLtoOpenStudio/resources/hvac.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (shorowit) [reply HPXMLtoOpenStudio/resources/hvac.rb:2856]: Will be addressed in the next commit.

File: HPXMLtoOpenStudio/resources/hvac.rb [hunk 6]
```diff
@@ -2849,52 +2850,74 @@ def self.process_neep_detailed_performance(hvac_system, mode, max_rated_fan_cfm,
...
+
+        high_odb_at_zero_power = calculate_odb_at_zero_power_or_capacity(data, user_odbs, :gross_input_power, true)
+        high_odb_at_zero_capacity = calculate_odb_at_zero_power_or_capacity(data, user_odbs, :gross_capacity, true)
+        outdoor_dry_bulbs << [high_odb_at_zero_power, high_odb_at_zero_capacity, 60.0].min # Max heating ODB
       end
+
+      # Extrapolate net capacity and input power at those outdoor drybulb temperatures per RESNET MINHERS Addendum 82.
       capacity_description = data[0].capacity_description
       outdoor_dry_bulbs.each do |target_odb|
         next if user_odbs.include? target_odb
```
Comment (shorowit) [inline HPXMLtoOpenStudio/resources/hvac.rb:2923]: Should really update this to check whether the target_odb is beyond the min (or max) of the user_odbs.

File: HPXMLtoOpenStudio/resources/hvac.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (shorowit) [reply HPXMLtoOpenStudio/resources/hvac.rb:3027]: We discussed and our solution is to disallow user-specified 60F heating and cooling datapoints. That solves part of the issue, but there is still a RESNET interpretation needed to resolve the remaining issue.

File: HPXMLtoOpenStudio/resources/hvac.rb [hunk 7]
```diff
@@ -2981,32 +3016,52 @@ def self.interpolate_to_odb_table_point(detailed_performance_data, capacity_desc
...
+    # Linearly extrapolate
     slope = (right_dp.send(property) - left_dp.send(property)) / (right_odb - left_odb)
     val = (target_odb - left_odb) * slope + left_dp.send(property)
+
+    if mode == :clg && target_odb < 82
+      # Ensure no less than 50% of the value at 82F
+      # FIXME: Need interpretation from RESNET on how this case should be handled
+      dp_82F = data.find { |dp| dp.outdoor_temperature == 82.0 }
+      val = [val, 0.5 * dp_82F.send(property)].max
+    elsif mode == :htg && target_odb > 47
```
[the remark this answers is not in the document]
Comment (shorowit) [reply HPXMLtoOpenStudio/resources/hvac.rb:3028]: In theory, we should be able to remove the (constant) 60F heating datapoint extrapolation because we've told EnergyPlus to use constant extrapolation beyond the TableLookup bounds. Test it.

File: Changelog.md [hunk 8]
```diff
@@ -4,6 +4,9 @@ __New Features__
 - HVAC modeling updates:
   - Updated DX heat pump and air conditioner models per RESNET MINHERS Addendum 82.
   - Allows optional EER or EER2 inputs for central air conditioners/heat pumps and mini-splits.
+  - **Breaking Change**: Detailed performance datapoints must now use `OutdoorTemperature` values of:
+    - Heating: 7F, 17F, and 5F (plus an optional 60F and an optional user-specified temperature <5F)
+    - Cooling: 82F and 95F (plus an optional 60F and an optional user-specified temperature >95F)
```
Comment (shorowit) [inline Changelog.md:9]: Need to update this too.

File: HPXMLtoOpenStudio/resources/hvac.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (shorowit) [reply HPXMLtoOpenStudio/resources/hvac.rb:3039]: I tested removing the creation of the 60F heating datapoint and indeed got the same answer. But the impact on the IDF is minimal and it's perhaps nice to be explicit, so I've decided to just keep the code how it is.

File: HPXMLtoOpenStudio/resources/hvac.rb [hunk 9]
```diff
@@ -2949,78 +2996,100 @@ def self.calculate_odb_at_zero_cop_or_capacity(data, user_odbs, property, find_h
...
-    target_dp = data.find { |dp| dp.outdoor_temperature == target_odb }
+  # @param datapoints [HPXML::CoolingDetailedPerformanceData or HPXML::HeatingDetailedPerformanceData] Array of detailed performance datapoints at a given speed
+  # @param capacity_description [String] The capacity description (HPXML::CapacityDescriptionXXX)
+  # @param target_property [Symbol] The datapoint property for the target value (e.g., :outdoor_temperature)
+  # @param target_value [Double] The target value to extrapolate to (F)
+  # @param property [Symbol] The datapoint property to extrapolate (e.g., :capacity, :efficiency_cop, etc.)
+  # @param slope_requirement [Symbol] The slope requirement (:positive or :negative)
+  # @return [Double] The extrapolated value (F)
+  def self.extrapolate_data_point(datapoints, capacity_description, target_property, target_value, property, slope_requirement = nil)
+    datapoints = datapoints.select { |dp| dp.capacity_description == capacity_description }
```
Comment (yzhou601) [inline HPXMLtoOpenStudio/resources/hvac.rb:3009]: Can we rename this variable to avoid modifying argument? Or is it intended?

Comment (shorowit) [reply HPXMLtoOpenStudio/resources/hvac.rb:3009]: It may look a little weird but I'd say it's intended. Any calls from `hvac.rb` already pass in datapoints that have been filtered by the CapacityDescription, but calls from `defaults.rb` and `hvac_sizing.rb` pass in the full datapoints list and need to be filtered. The code in this method should always operate on the filtered list of datapoints. Open to suggestions though...

Comment (shorowit) [reply HPXMLtoOpenStudio/resources/hvac.rb:3009]: Oh wait, I think I understand your concern now. It doesn't actually modify the array that the calling method passed in.

**Review by yzhou601 — APPROVED** [review]
This looks good to me. Thanks @shorowit for all the improvements here!
